Added support for JASC PaintShop Pro and GIMP color palettes

Mailaender opened the pull request:
Closes https://github.com/OpenRA/OpenRA/issues/11385.

[PaintShopPro.pal](https://raw.githubusercontent.com/Mailaender/ArtSrc/f3e8d2f33a34e1ce3258b3b8dd21543acd72a2d4/ra/palettes/PaintShopPro.pal) is the RA95 palette in JASC format for everyone who wants to test.

Mailaender committed:
Add support for JASC PaintShop Pro .pal files.

Mailaender committed:
Add support for GIMP color palettes.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/PaletteFromGimpFile.cs:
```diff
@@ -0,0 +1,110 @@
+			{
+				var lines = s.ReadAllLines().GetEnumerator();
+
+				lines.MoveNext();
```
Will this throw on an empty file?

Mailaender replied:
No.

RoosterDragon replied:
But `line.Current` might if `lines.MoveNext()` returns false.

You probably want 
```c#
if (!lines.MoveNext() || lines.Current != "GIMP Palette")
    throw ...
```

and similarly in the other file.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/PaletteFromGimpFile.cs:
```diff
@@ -0,0 +1,110 @@
+					if (!string.IsNullOrEmpty(lines.Current) && char.IsDigit(lines.Current[0]))
+					{
+						var rgb = lines.Current.Split((char[]) null, StringSplitOptions.RemoveEmptyEntries);
+						if (rgb.Length < 3)
```
This should check for `!= 3`.

Mailaender replied:
My example files from http://www.davidrevoy.com/article27/ come with the name of the color after the RGB, so 4 would be valid as well. I will rename the variable instead.

Phrohdoh commented on OpenRA.Mods.Common/Traits/World/PaletteFromGimpFile.cs:
```diff
@@ -0,0 +1,110 @@
+				while (lines.MoveNext() && i < Palette.Size)
+				{
+					// Skip until first color and ignore comments.
+					if (!string.IsNullOrEmpty(lines.Current) && char.IsDigit(lines.Current[0]))
```
A triplet with leading whitespace is valid.
Split on whitespace and trim all, then check the length.

pchote replied:
We have FieldLoader for this.  It can parse directly to `char[]` or throw an error on invalid input string.

Phrohdoh replied:
This branch may be helpful: https://github.com/Phrohdoh/OpenRA/commit/e60540975977c470242496650696f773a846f330

Mailaender replied:
Updated with a `Trim()` to not ignore the leading whitespace colors.

chrisforbes commented on OpenRA.Mods.Common/Traits/World/PaletteFromGimpFile.cs:
```diff
@@ -0,0 +1,110 @@
+		public readonly string Filename = null;
+
+		[Desc("Map listed indices to shadow. Ignores previous color.")]
+		public readonly int[] ShadowIndex = { };
```
This doesnt work -- you should be passing it to the ImmutablePalette ctor.

Mailaender replied:
Thanks. I noticed during testing, but wasn't sure if the file was butchered. Should be fixed now.

RoosterDragon commented on OpenRA.Game/Graphics/Palette.cs:
```diff
@@ -144,6 +144,16 @@ public ImmutablePalette(IEnumerable<uint> sourceColors)
 				colors[i++] = sourceColor;
 		}
+
+		public ImmutablePalette(IEnumerable<uint> sourceColors, int[] remapShadow)
```
Can call `:this(sourceColors)` here to save some code repetition.

Mailaender replied:
Good idea. Thanks.

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/PaletteFromJascFile.cs:
```diff
@@ -0,0 +1,111 @@
+					throw new InvalidDataException("File {0} is not a valid JASC palette.".F(info.Filename));
+
+				lines.MoveNext();
+				lines.MoveNext();
```
Consider throwing if either of these fails.

pchote commented:
Can you please provide a gimp palette file to test with?

RoosterDragon commented on OpenRA.Mods.Common/Traits/World/PaletteFromGimpFile.cs:
```diff
@@ -0,0 +1,110 @@
+					if (!int.TryParse(rgbPlusColorName[1], out g))
+						throw new InvalidDataException("Invalid G value: {0}".F(rgbPlusColorName[1]));
+
+					if (!int.TryParse(rgbPlusColorName[2], out b))
```
Don't we actually want `byte.TryParse` for all of these? `Color.FromArgb` accepts ints for ease of use, but actually only accepts values of 0-255.

Mailaender committed:
Add a JASC PaintShopPro example file.

Mailaender committed:
Add example GIMP palettes

Mailaender commented:
> Can you please provide a gimp palette file to test with?

Added palettes from:
http://www.sleipnirstuff.com/forum/viewtopic.php?f=83&t=16176
https://cnc-comm.com/command-and-conquer/downloads/modding-tools/purple-palettes

and updated the code as suggested by @RoosterDragon.

penev92 commented on OpenRA.Mods.Common/Traits/World/PaletteFromJascFile.cs:
```diff
@@ -0,0 +1,106 @@
+			var colors = new uint[Palette.Size];
+			using (var s = world.Map.Open(info.Filename))
+			{
+				var lines = s.ReadAllLines().GetEnumerator();
```
This needs an `using` statement.

penev92 commented on OpenRA.Mods.Common/Traits/World/PaletteFromJascFile.cs:
```diff
@@ -0,0 +1,106 @@
+				var lines = s.ReadAllLines().GetEnumerator();
+
+				if (!lines.MoveNext() || lines.Current != "JASC-PAL" || !lines.MoveNext() || !lines.MoveNext())
+					throw new InvalidDataException("File {0} is not a valid JASC palette.".F(info.Filename));
```
```"File `{0}` is not...```, please.

penev92 commented on OpenRA.Mods.Common/Traits/World/PaletteFromJascFile.cs:
```diff
@@ -0,0 +1,106 @@
+				if (!lines.MoveNext() || lines.Current != "JASC-PAL" || !lines.MoveNext() || !lines.MoveNext())
+					throw new InvalidDataException("File {0} is not a valid JASC palette.".F(info.Filename));
+
+				if (int.Parse(lines.Current) != Palette.Size)
```
`lines.Current` could use a null check here.

Phrohdoh replied:
Prefer `TryParse` variants over `Parse` unless you have already checked the data.

penev92 commented on OpenRA.Mods.Common/Traits/World/PaletteFromJascFile.cs:
```diff
@@ -0,0 +1,106 @@
+				if (int.Parse(lines.Current) != Palette.Size)
+					throw new InvalidDataException("Only {0} color palettes are supported.".F(Palette.Size));
+
+				byte r, g, b = 0;
```
Style nit: ` = 0` is redundant. Also these could be separately declared where they are used, but that's a matter of preference.

penev92 commented on OpenRA.Mods.Common/Traits/World/PaletteFromJascFile.cs:
```diff
@@ -0,0 +1,106 @@
+
+				while (lines.MoveNext() && i < Palette.Size)
+				{
+					var rgb = lines.Current.Split(' ');
```
`lines.Current` could use a null check here.

penev92 reviewed (commented):
I left a bunch of simple stylistic comments, but 👍  once those are cleared up.

penev92 commented on OpenRA.Mods.Common/Traits/World/PaletteFromJascFile.cs:
```diff
@@ -0,0 +1,106 @@
+				{
+					var rgb = lines.Current.Split(' ');
+					if (rgb.Length != 3)
+						throw new InvalidDataException("Invalid RGB triplet: {0}".F(rgb));
```
This should be `"Invalid RGB triplet: ({0})".F(string.Join(" ", rgb))`.

penev92 commented:
It has been a while, and while I'd prefer @Mailaender came back to finish this, at some point I may have to push the requested fixes myself.
I suspect I'll have time for that early next week.

Mailaender commented:
Updated it myself.

pchote commented on OpenRA.Mods.Common/Traits/World/PaletteFromFile.cs:
```diff
@@ -15,6 +15,7 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
+	[Desc("Load Westwood .pal files.")]
```
This format isn't Westwood specific.  It directly maps to the VGA palette registers.

Mailaender replied:
Renamed.

pchote reviewed (changes requested):
These file formats and the parsers are virtually identical, with only minor differences in the file headers.  Can you please combine these into a single trait?

Mailaender commented:
> These file formats and the parsers are virtually identical, with only minor differences in the file headers. Can you please combine these into a single trait?

If that is mandatory, then someone else should take over. I already did enough requested changes here.

Arular101 commented:
Please don't forget to update the copyright notice year.

reaperrr commented:
> If that is mandatory, then someone else should take over. I already did enough requested changes here.

Filed #14851.
